Update Syncromatics API, add per-vehicle timestamp, remove TripUpdates

barbeau opened the pull request:
This is a continuation of https://github.com/CUTR-at-USF/bullrunner-gtfs-realtime-generator/pull/24, so @minhhpham and I can both push to the same branch.

Commit:
Version 1.0.0
1. Change data API
2. Create API key file not included in source code
3. Update GTFS file with new bullrunner webpage link and new route (Lot 18 Express)
4. Update README

Commit:
Version 1.0.0-SNAPSHOT

1. Change data API
2. Create API key file not included in source code
3. Update GTFS file with new bullrunner webpage link and new route (Lot 18 Express)
4. Update README

Commit:
Versino 1.0.0-SNAPSHOT

1. Changes to GTFS: removed Lot 18 Express Route, added MSC Express to routes, frequencies, shapes, trips, strop_times. Added direction_id = 0 to all trips
2. The program can be run from either the main directory or the target directory, it will look for GTFS and key files from current directory or parent directory. Will display error messages asking for these 2 if it can't find files and then exit.
3. Cap bearing and speed at 2 decimal places
4. Change the run instruction in README. Instruction shows to run from main directory (even though it can also run from target directory), changed to port 8088
5. Removed code related to TripUpdates

Commit:
Some updates
1. Added Javadocs
2. Added back the removed inject methods
3. Update GTFS files: renamed route_long_name and revised stop_times and headway_secs for MSC Express, added calendar_dates.txt

Commit:
added trip_id and schedule_relationship

barbeau commented:
@minhhpham You should be able to check out this branch locally by doing:

~~~
git remote add upstream git@github.com:CUTR-at-USF/bullrunner-gtfs-realtime-generator.git
git fetch upstream
git checkout -b update-syncromatics-api upstream/update-syncromatics-api
git merge master
~~~

Then, after I add commits, you can bring them into your repo by doing:

~~~
git pull upstream update-syncromatics-api
~~~

And, you can commit to your repo and push to this PR too.

I'm just going to do some code cleanup and then will have you review.

barbeau commented on src/main/java/edu/usf/cutr/gtfs_realtime/bullrunner/GtfsRealtimeProviderImpl.java:
```diff
@@ -16,688 +16,446 @@
+
+            // check if pattern name matches
+            if (pattern_name.equals(Pattern_name_C)){
+                if (route_id.equals(Route_id_C))
```
@minhhpham I think `if (route_id.equals(Route_id_C))` is always true?  I'm going to remove it in my next commit, but please let me know if you think it needs to stay for some reason.

barbeau replied:
Actually, looking more the value gets passed in from the JSON response, so on 2nd thought we do need it.  Let me know if I'm wrong.

barbeau commented on src/main/java/edu/usf/cutr/gtfs_realtime/bullrunner/GtfsRealtimeProviderImpl.java:
```diff
@@ -16,688 +16,446 @@
+                if (route_id.equals(Route_id_C))
+                    response_out.put(vehicleObj);
+            } else if (pattern_name.equals(Pattern_name_MSC)){
+                if (route_id.equals(Route_id_MSC))
```
@minhhpham I think `if (route_id.equals(Route_id_MSC))` is always true?  I'm going to remove it in my next commit, but please let me know if you think it needs to stay for some reason.

barbeau replied:
Actually, looking more the value gets passed in from the JSON response, so on 2nd thought we do need it.  Let me know if I'm wrong.

minhhpham replied:
We need this to assign vehicles to the correct `route_id`. `refreshTripVehicle()` loops through `route_id` in GTFS files and request vehicle info from the API. Since the API returns both route C and MSC Express vehicle as a response to the request for route C, these 4 lines attempt to match `pattern_name` with `route_id`.

barbeau committed:
Code cleanup

* Refactor variables for typical Java code conventions
* Reformat code and white space
* Narrow scope of exceptions thrown
* Clarify comments and method Javadocs
* Add braces for all IF/WHILEs
* Add/update license headers
* Log request URLs

barbeau commented:
@minhhpham As we discussed via Slack I'm still seeing this exception when running the feed:

~~~
2018-09-18 12:24:47,325 INFO  [Server.java:284] : jetty-9.0.5.v20130815
2018-09-18 12:24:47,496 INFO  [AbstractConnector.java:265] : Started ServerConnector@add0edd{HTTP/1.1}{0.0.0.0:8088}
2018-09-18 12:24:48,325 INFO  [GtfsRealtimeProviderImpl.java:186] : starting GTFS-realtime service
2018-09-18 12:24:48,330 INFO  [GtfsRealtimeProviderImpl.java:439] : Refreshing vehicles...
2018-09-18 12:24:48,861 INFO  [GtfsRealtimeProviderImpl.java:371] : http://api.syncromatics.com/portal/v1/vehicles/483/arrivals?count=100&api-key=ad0b0a576fa1e506fd47eeb077c5076559981402fd127afe183492395fa0ef2a
2018-09-18 12:24:49,141 INFO  [GtfsRealtimeProviderImpl.java:371] : http://api.syncromatics.com/portal/v1/vehicles/757/arrivals?count=100&api-key=ad0b0a576fa1e506fd47eeb077c5076559981402fd127afe183492395fa0ef2a
2018-09-18 12:24:49,376 INFO  [GtfsRealtimeProviderImpl.java:371] : http://api.syncromatics.com/portal/v1/vehicles/2710/arrivals?count=100&api-key=ad0b0a576fa1e506fd47eeb077c5076559981402fd127afe183492395fa0ef2a
2018-09-18 12:24:49,561 WARN  [GtfsRealtimeProviderImpl.java:442] : Error in vehicle refresh task
org.json.JSONException: JSONArray[0] not found.
	at org.json.JSONArray.get(JSONArray.java:234)
	at org.json.JSONArray.getJSONObject(JSONArray.java:326)
	at edu.usf.cutr.gtfs_realtime.bullrunner.GtfsRealtimeProviderImpl.downloadVehiclesRouteC(GtfsRealtimeProviderImpl.java:395)
	at edu.usf.cutr.gtfs_realtime.bullrunner.GtfsRealtimeProviderImpl.refreshTripVehicle(GtfsRealtimeProviderImpl.java:216)
	at edu.usf.cutr.gtfs_realtime.bullrunner.GtfsRealtimeProviderImpl.access$200(GtfsRealtimeProviderImpl.java:60)
	at edu.usf.cutr.gtfs_realtime.bullrunner.GtfsRealtimeProviderImpl$VehiclesRefreshTask.run(GtfsRealtimeProviderImpl.java:440)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
~~~

Commit:
Fixed bug #27
If the API response is empty, we simply cannot determine whether a bus is of route C or MSC Express. In this case, we still export the vehicle location to the feed, but not provide route_id or trip_id

barbeau commented on src/main/java/edu/usf/cutr/gtfs_realtime/bullrunner/GtfsRealtimeProviderImpl.java:
```diff
@@ -219,17 +221,31 @@ private void refreshTripVehicle() throws IOException, JSONException {
+            HashSet addedVehicleIds = new HashSet(); // keep track of vehicles that have been added to the feed
             for (int k = 0; k < vehicleArray.length(); k++) {
                 JSONObject vehicleObj = vehicleArray.getJSONObject(k);
+                // check if vehicle has been added to feed
```
Under what conditions would a duplicate vehicle get added to the feed if we don't use the HashSet?

From our previous discussion I thought this was intended to solve the unknown route ID issue.  But, if we're checking for "unknown" route_ids in the below if statement, do we actually need the HashSet to make sure we're not adding duplicate vehicles to the feed?

Let me know if the duplicate vehicle IDs issue is separate from the unknown route issue.

minhhpham replied:
Both issues happen at the same time. The program loop through `route_id`s given in GTFS. Syncromatics data has the same `route_id` for route C and MSC Express. When we can't figure out whether a bus is of route C or MSC Express, it is picked up by both routes and hence duplicated in the feed.

barbeau replied:
So do we need to call `downloadVehiclesRouteC()` twice then, once for each of our GTFS `route_ids`?

It seems like we're duplicating the exact same calls to the server and looping through the same responses twice for external_route_id 156. It seems like it would be simpler to just call `downloadVehiclesRouteC()` once and then within that method call the arrivals APIs, and then create a single `JSONArray` with the route_id for the vehicles set to "C", "MSC Express", or "Unknown".  It would half the number of API calls for "C" and "MSC Express", unless I'm missing something.  This would also let you get rid of the HashSet later to remove the duplicates.

What do you think of the above?

minhhpham replied:
Yes that works. That's a little more to code but will reduce run time. I will change the methods that way.

barbeau replied:
Ok, great. I suppose we may still need a Hash Set before calling this method just to keep track if we've already called the downloadVehiclesRouteC() method for the given external_route_id when iterating through the routes from routes.txt. This is an extensible way to set it up so that we could use the same design for the Lot 18 Express later if that's part of Route A, and we add another record to routes.txt later for that route.

minhhpham replied:
I used `if (route_id.equals("MSC Express")) continue;` for this idea.

barbeau replied:
@minhhpham I guess that works fine too, and is simpler. :+1:

barbeau commented on src/main/java/edu/usf/cutr/gtfs_realtime/bullrunner/GtfsRealtimeProviderImpl.java:
```diff
@@ -246,6 +262,7 @@ private void refreshTripVehicle() throws IOException, JSONException {
                 vpBuilder.setTimestamp(
                         Instant.parse(vehicleObj.getString(FIELD_LAST_UPDATED)).getEpochSecond()
                 );
+                vpBuilder.hasVehicle();
```
I believe this line of code can be removed - it doesn't appear to do anything.

minhhpham replied:
I added this by mistake.

barbeau commented on src/main/java/edu/usf/cutr/gtfs_realtime/bullrunner/GtfsRealtimeProviderImpl.java:
```diff
@@ -16,688 +16,457 @@
+                        Instant.parse(vehicleObj.getString(FIELD_LAST_UPDATED)).getEpochSecond()
+                );
+                vpBuilder.hasVehicle();
+                if (vehicleObj.getDouble(FIELD_PASSENGER_LOAD) <= 0) {
```
@minhhpham Could you take a look at the passenger load values?  It looks like we're always reporting `MANY_SEATS_AVAILABLE`.  I think the value returned from the API may have changed from an integer (which the below IF statements are written for) to a fraction - it looks like this in the new API - `"passengerLoad": 0.11`.

Commit:
Fixed Occupancy_Status, made API calls more efficient

barbeau committed:
Fix Javadocs, add debug logging for arrival API

barbeau committed:
Code cleanup

* Fix whitespace/formatting, remove unused method

barbeau commented:
@minhhpham This is looking great!  Thanks for all your work on this. Do you have any other changes, or can I squash and merge this?

minhhpham commented:
Looks good to me too. I don't have any other change.

barbeau commented:
Thanks @minhhpham!
